tools/pubsys/src/repo/validate_repo/mod.rs

```diff
@@ -0,0 +1,252 @@
+//! The validate_repo module owns the 'validate-ami' subcommand and provides methods for validating
+//! a given TUF repository by attempting to loadi the repository and downloading its targets.
+
+use super::RepoTransport;
+use crate::repo::{error as repo_error, repo_urls};
+use crate::Args;
+use futures::future::join_all;
+use log::{error, info, trace};
+use pubsys_config::InfraConfig;
+use rand::seq::SliceRandom;
+use snafu::{ensure, OptionExt, ResultExt};
+use std::fs::File;
+use std::io;
+use std::io::Read;
+use std::path::PathBuf;
+use structopt::StructOpt;
+use tempfile::tempdir;
+use tough::{ExpirationEnforcement, Limits, Repository, Settings};
+use url::Url;
+
+/// Validates a set of TUF repositories
+#[derive(Debug, StructOpt)]
+#[structopt(setting = clap::AppSettings::DeriveDisplayOrder)]
+pub(crate) struct ValidateRepoArgs {
+    #[structopt(long)]
+    /// Use this named repo infrastructure from Infra.toml
+    repo: String,
+
+    #[structopt(long)]
+    /// The architecture of the repo being validated
+    arch: String,
+    #[structopt(long)]
+    /// The variant of the repo being validated
+    variant: String,
+
+    #[structopt(long, parse(from_os_str))]
+    /// Path to root.json for this repo
+    root_role_path: PathBuf,
+
+    #[structopt(default_value = "100", long)]
+    /// Randomly samples and retrieves specified percentage of targets
+    percent_target_files: u8,
+}
+
+fn attempt_read_target<T>(
+    repo: &Repository<'_, T>,
+    target: String,
+) -> Result<Option<impl Read>, repo_error::Error>
+where
+    T: tough::Transport,
+{
+    let max_attempts = 5;
+    for attempt in 1..=max_attempts {
+        match repo.read_target(&target) {
+            Err(err) => match err {
+                // Retry if it's a transport failure, might be due to intermittent connection issues
+                tough::error::Error::Transport { .. } => {
+                    error!(
+                        "Repo target fetch failure on attempt {} out of {}",
+                        attempt, max_attempts
+                    );
+                    if attempt >= max_attempts {
+                        return Err(repo_error::Error::ReadTarget {
+                            target,
+                            source: err,
+                        });
+                    }
+                }
+                // For other errors we should just return immediately and not retry
+                err => {
+                    return Err(repo_error::Error::ReadTarget {
+                        target,
+                        source: err,
+                    })
+                }
+            },
+            Ok(target_reader) => return Ok(target_reader),
+        };
+    }
+
+    Ok(None)
+}
+
+async fn attempt_download_target(
```

Comment:
Hmm, why does this need to be async now?  The implementation didn't change, and it seems synchronous.

Reply:
You're right, this doesn't need to be async. Removing..